README.md
polkit-kde-agent is based on polkit-kde-agent [gitlab](https://invent.kde.org/plasma/polkit-kde-agent-1) 


## Features
 
* Virtual secure keyboard

* Run on JingOS platform

* Brand new UI & UE with JingOS-style , based on JingUI Framework

* Support keyboard & touchpad & mouse & screen touch

* All keys support pressed / hovered effects

* Well-designed interface material:

  * Font
  * Icon
  * Picture
  


## Links

* Home page: https://www.jingos.com/

* Project page: https://invent.kde.org/jingosdev/polkit-kde-agent-1

* Issues: https://invent.kde.org/jingosdev/polkit-kde-agent-1/issues

* Development channel: https://forum.jingos.com/



## Dependencies

* Qt5 

* Cmake

* KI18n

* Kirigami (JingOS Version)

* DBusAddons

* PolkitQt5-1

* PlasmaQuick



## Build

To build polikit-kde-agent from source on Linux, execute the below commands.



### Compile

```sh
cd polikit-kde-agent
mkdir build
cd build
cmake ..
make -j$(nproc)
```



#### Run

```
bin/polikit-kde-agent
```



#### Install

```
sudo make install
```



debian/changelog
polkit-kde-agent-1 (4:5.20.5) focal; urgency=medium

  * Format version

 -- JingOS CI <jingos@jingos.com>  Nov, 17 Wed 2021 21:53:20 +0800

polkit-kde-agent-1 (4:5.20.4-0neon) focal; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 01 Dec 2020 15:20:59 +0000

polkit-kde-agent-1 (4:5.20.3-0neon) focal; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 10 Nov 2020 21:16:41 +0000

polkit-kde-agent-1 (4:5.20.2-0neon) focal; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 27 Oct 2020 16:31:02 +0000

polkit-kde-agent-1 (4:5.20.1-0neon) focal; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 20 Oct 2020 13:09:00 +0000

polkit-kde-agent-1 (4:5.20.0-0neon) focal; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Mon, 12 Oct 2020 16:43:24 +0000

polkit-kde-agent-1 (4:5.19.5-0neon) focal; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 01 Sep 2020 18:24:30 +0000

polkit-kde-agent-1 (4:5.19.4-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 28 Jul 2020 14:17:20 +0000

polkit-kde-agent-1 (4:5.19.3-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 07 Jul 2020 13:29:17 +0000

polkit-kde-agent-1 (4:5.19.2-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 23 Jun 2020 13:30:04 +0000

polkit-kde-agent-1 (4:5.19.1-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 16 Jun 2020 14:22:13 +0000

polkit-kde-agent-1 (4:5.19.0-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Mon, 08 Jun 2020 11:36:33 +0000

polkit-kde-agent-1 (4:5.18.5-0neon) bionic; urgency=medium

  [ Jonathan Riddell ]
  * new release

  [ Neon CI ]
  * New release

 -- Neon CI <neon@kde.org>  Tue, 05 May 2020 16:19:04 +0000

polkit-kde-agent-1 (4:5.18.3-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 10 Mar 2020 14:51:19 +0000

polkit-kde-agent-1 (4:5.18.2-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 25 Feb 2020 17:20:57 +0000

polkit-kde-agent-1 (4:5.18.1-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Wed, 19 Feb 2020 13:09:46 +0000

polkit-kde-agent-1 (4:5.18.0-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Sun, 09 Feb 2020 22:39:53 +0000

polkit-kde-agent-1 (4:5.17.5-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Wed, 08 Jan 2020 13:40:34 +0000

polkit-kde-agent-1 (4:5.17.4-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 03 Dec 2019 15:17:01 +0000

polkit-kde-agent-1 (4:5.17.3-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 12 Nov 2019 13:51:48 +0000

polkit-kde-agent-1 (4:5.17.2-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 29 Oct 2019 22:36:41 +0000

polkit-kde-agent-1 (4:5.17.1-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 22 Oct 2019 21:31:21 +0000

polkit-kde-agent-1 (4:5.17.0-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Mon, 14 Oct 2019 15:45:24 +0000

polkit-kde-agent-1 (4:5.16.5-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 03 Sep 2019 14:46:06 +0000

polkit-kde-agent-1 (4:5.16.4-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 30 Jul 2019 12:26:54 +0000

polkit-kde-agent-1 (4:5.16.3-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 09 Jul 2019 17:14:21 +0000

polkit-kde-agent-1 (4:5.16.2-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 25 Jun 2019 14:13:34 +0000

polkit-kde-agent-1 (4:5.16.1-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 18 Jun 2019 13:29:12 +0000

polkit-kde-agent-1 (4:5.16.0-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Mon, 10 Jun 2019 09:22:27 +0000

polkit-kde-agent-1 (4:5.15.5-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 07 May 2019 12:25:16 +0000

polkit-kde-agent-1 (4:5.15.4-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 02 Apr 2019 13:00:00 +0000

polkit-kde-agent-1 (4:5.15.3-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 12 Mar 2019 12:15:51 +0000

polkit-kde-agent-1 (4:5.15.2-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 26 Feb 2019 16:22:23 +0000

polkit-kde-agent-1 (4:5.15.1-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Wed, 20 Feb 2019 14:44:35 +0000

polkit-kde-agent-1 (4:5.15.0-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Mon, 11 Feb 2019 14:25:18 +0000

polkit-kde-agent-1 (4:5.14.5-0neon) bionic; urgency=medium

  * New release

 -- Neon CI <neon@kde.org>  Tue, 08 Jan 2019 13:20:13 +0000

polkit-kde-agent-1 (4:5.14.5-1~) UNRELEASED; urgency=medium

  * New upstream release (5.14.5).

 -- Maximiliano Curia <maxy@debian.org>  Fri, 11 Jan 2019 16:12:12 -0300

polkit-kde-agent-1 (4:5.14.3-1) unstable; urgency=medium

  * Update upsteam signing-key
  * New upstream release (5.14.3).
  * Update build-deps and deps with the info from cmake
  * Bump group breaks (4:5.14)
  * Release to unstable

 -- Maximiliano Curia <maxy@debian.org>  Fri, 23 Nov 2018 08:51:11 -0300

polkit-kde-agent-1 (4:5.13.4-1) unstable; urgency=medium

  * New upstream release (5.13.4).
  * Release to unstable

 -- Maximiliano Curia <maxy@debian.org>  Sun, 19 Aug 2018 23:18:24 +0200

polkit-kde-agent-1 (4:5.13.1-1) unstable; urgency=medium

  * New upstream release (5.13.1).
  * Update build-deps and deps with the info from cmake
  * Bump group breaks (4:5.13)
  * Release to unstable

 -- Maximiliano Curia <maxy@debian.org>  Tue, 26 Jun 2018 13:43:27 +0200


main.cpp
/*  This file is part of the KDE project
    Copyright (C) 2009 Jaroslav Reznik <jreznik@redhat.com>

    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Library General Public License for more details.

    You should have received a copy of the GNU Library General Public License
    along with this library; see the file COPYING.LIB.  If not, write to
    the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
    Boston, MA 02110-1301, USA.

*/

#include "policykitlistener.h"
#include "config.h"
// KF
#include <KAboutData>
#include <KLocalizedString>
#include <KCrash>
#include <KDBusService>
#include <KConfig>
#include <KLocalizedContext>
// PolkitQt1
#include <PolkitQt1/Subject>
// Qt
#include <QApplication>
#include <QSessionManager>
#include <QDebug>
// std
#if HAVE_SYS_PRCTL_H
#include <sys/prctl.h>
#endif
#if HAVE_SYS_PROCCTL_H
#include <unistd.h>
#include <sys/procctl.h>
#endif


int main(int argc, char *argv[])
{
    // disable ptrace
#if HAVE_PR_SET_DUMPABLE
    prctl(PR_SET_DUMPABLE, 0);
#endif
#if HAVE_PROC_TRACE_CTL
    int mode = PROC_TRACE_CTL_DISABLE;
    procctl(P_PID, getpid(), PROC_TRACE_CTL, &mode);
#endif

    QCoreApplication::setAttribute(Qt::AA_UseHighDpiPixmaps, true);
    
    KCrash::setFlags(KCrash::AutoRestart);

    QApplication app(argc, argv);

    app.setQuitOnLastWindowClosed(false);
    
    KLocalizedString::setApplicationDomain("polkit-kde-authentication-agent-1");

    KAboutData aboutData("polkit-kde-authentication-agent-1", i18n("PolicyKit1 KDE Agent"), POLKIT_KDE_1_VERSION);
    aboutData.addLicense(KAboutLicense::GPL);
    aboutData.addCredit(i18n("(c) 2009 Red Hat, Inc."));
    aboutData.addAuthor(i18n("Lukáš Tinkl"), i18n("Maintainer"), "ltinkl@redhat.com");
    aboutData.addAuthor(i18n("Jaroslav Reznik"), i18n("Former maintainer"), "jreznik@redhat.com");
    aboutData.setProductName("policykit-kde/polkit-kde-authentication-agent-1");

    KAboutData::setApplicationData(aboutData);

    // ensure singleton run
    KDBusService service(KDBusService::Unique);

    // disable session management
    auto disableSessionManagement = [](QSessionManager &sm) {
        sm.setRestartHint(QSessionManager::RestartNever);
    };

    QObject::connect(&app, &QGuiApplication::commitDataRequest, disableSessionManagement);
    QObject::connect(&app, &QGuiApplication::saveStateRequest, disableSessionManagement);

    // register agent
    PolicyKitListener *listener = new PolicyKitListener(&app);

    PolkitQt1::UnixSessionSubject session(getpid());

    const bool result = listener->registerListener(session, "/org/kde/PolicyKit1/AuthenticationAgent");

    qDebug() << "Authentication agent result:" << result;

    if (!result) {
        qWarning() << "Couldn't register listener!";
        exit(1);
    }

    app.exec();
}



policykitlistener.h
#ifndef POLICYKITLISTENER_H
#define POLICYKITLISTENER_H

/*  This file is part of the KDE project
    Copyright (C) 2009 Jaroslav Reznik <jreznik@redhat.com>

    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Library General Public License for more details.

    You should have received a copy of the GNU Library General Public License
    along with this library; see the file COPYING.LIB.  If not, write to
    the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
    Boston, MA 02110-1301, USA.
*/

#include <QPointer>
#include <QHash>
#include <QQuickView>
#include <QStringList>
#include <QObject>
#include <PolkitQt1/Agent/Listener>
#include <QString>

using namespace PolkitQt1::Agent;

class PolicyKitListener : public Listener
{
    Q_OBJECT
    Q_CLASSINFO("D-Bus Interface", "org.kde.Polkit1AuthAgent")

public:
    explicit PolicyKitListener(QObject *parent = nullptr);
    ~PolicyKitListener() override;

    Q_INVOKABLE void dialogAccepted( const QString &passwd );

    Q_INVOKABLE QStringList getIdentities() const;
    //add by huanlele
    Q_INVOKABLE void invokeSendCancelSig();
    Q_INVOKABLE void invokeSendConfirmSig();
    Q_INVOKABLE void invokeSendErrorSig(int errorCode);

    Q_INVOKABLE QString getMessage();

    QString  errorMessage();

    Q_INVOKABLE void dialogCanceled();

signals:
    void sigCancel();
    void sigConfirm();
    void sigError(int errorCode);
    //end by huanlele
    void errorMessageChanged(QString errorStr);


public slots:
    void initiateAuthentication(const QString &actionId,
                                const QString &message,
                                const QString &iconName,
                                const PolkitQt1::Details &details,
                                const QString &cookie,
                                const PolkitQt1::Identity::List &identities,
                                PolkitQt1::Agent::AsyncResult* result) override;
    bool initiateAuthenticationFinish() override;
    void cancelAuthentication() override;
    void tryAgain();
    void finishObtainPrivilege();
    void completed(bool gainedAuthorization);
    void setWIdForAction(const QString &action, qulonglong wID);

private:
    QPointer<QQuickView> m_view;
    QPointer<Session> m_session;
    bool m_inProgress;
    bool m_gainedAuthorization;
    bool m_wasCancelled;
    int m_numTries;
    PolkitQt1::Identity::List m_identities;
    PolkitQt1::Agent::AsyncResult* m_result;
    QString m_cookie;
    PolkitQt1::Identity m_selectedUser;
    QHash< QString, qulonglong > m_actionsToWID;
    QString m_error;
    QString m_message;

    void setErrorMessage(const QString &errorMessage);

private slots:
    void userSelected(const PolkitQt1::Identity &identity);
    void userCancel();

};

#endif


policykitlistener.cpp
/*  This file is part of the KDE project
    Copyright (C) 2009 Jaroslav Reznik <jreznik@redhat.com>

    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Library General Public License for more details.

    You should have received a copy of the GNU Library General Public License
    along with this library; see the file COPYING.LIB.  If not, write to
    the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
    Boston, MA 02110-1301, USA.
*/

#include <QDBusConnection>
#include <QDebug>
#include <QQuickView>
#include <QQmlContext>
#include <QStringList>
#include <QQuickWindow>

#include <KWindowSystem>
// KF
#include <KAboutData>
#include <KLocalizedString>
#include <KCrash>
#include <KDBusService>
#include <KConfig>
#include <KLocalizedContext>


#include <PolkitQt1/Agent/Session>
#include <PolkitQt1/Subject>
#include <PolkitQt1/Identity>
#include <PolkitQt1/Details>

#include "policykitlistener.h"
#include "polkit1authagentadaptor.h"


PolicyKitListener::PolicyKitListener(QObject *parent)
        : Listener(parent)
        , m_inProgress(false)
        , m_selectedUser(nullptr)
{
    (void) new Polkit1AuthAgentAdaptor(this);

    if (!QDBusConnection::sessionBus().registerObject("/org/kde/Polkit1AuthAgent", this,
                                                     QDBusConnection::ExportScriptableSlots |
                                                     QDBusConnection::ExportAllSignals |      //add by huanlele
                                                     QDBusConnection::ExportScriptableProperties |
                                                     QDBusConnection::ExportAdaptors))
    {
        qWarning() << "Could not initiate DBus helper!";
    }

    qDebug() << "Listener online";
}

PolicyKitListener::~PolicyKitListener()
{

}

//add by huanlele
void PolicyKitListener::invokeSendCancelSig()
{
    QDBusMessage message = QDBusMessage::createSignal("/org/kde/Polkit1AuthAgent", "org.kde.Polkit1AuthAgent", "sigCancel");
    bool rv = QDBusConnection::sessionBus().send(message);
    if(rv == false) {
        qWarning() << Q_FUNC_INFO << " send dbus signal fail  path is /org/kde/Polkit1AuthAgent "
            << " interface is org.kde.Polkit1AuthAgent  signal is sigCancel";
    }
}

void PolicyKitListener::invokeSendConfirmSig()
{
    QDBusMessage message =QDBusMessage::createSignal("/org/kde/Polkit1AuthAgent", "org.kde.Polkit1AuthAgent", "sigConfirm");
    bool rv = QDBusConnection::sessionBus().send(message);
    if(rv == false) {
        qWarning() << Q_FUNC_INFO << " send dbus signal fail  path is /org/kde/Polkit1AuthAgent "
            << " interface is org.kde.Polkit1AuthAgent  signal is sigConfirm";
    }
}

void PolicyKitListener::invokeSendErrorSig(int errorCode)
{
    QDBusMessage message =QDBusMessage::createSignal("/org/kde/Polkit1AuthAgent", "org.kde.Polkit1AuthAgent", "sigError");
    message << errorCode;
    bool rv = QDBusConnection::sessionBus().send(message);
    if(rv == false) {
        qWarning() << Q_FUNC_INFO << " send dbus signal fail  path is /org/kde/Polkit1AuthAgent "
            << " interface is org.kde.Polkit1AuthAgent  signal is sigError";
    }
}
// end by huanlele
void PolicyKitListener::setWIdForAction(const QString& action, qulonglong wID)
{
    qDebug() << "On to the handshake";
    m_actionsToWID[action] = wID;
}

void PolicyKitListener::initiateAuthentication(const QString &actionId,
        const QString &message,
        const QString &iconName,
        const PolkitQt1::Details &details,
        const QString &cookie,
        const PolkitQt1::Identity::List &identities,
        PolkitQt1::Agent::AsyncResult* result)
{
    qDebug() << "Initiating authentication";

    if (m_inProgress) {
        result->setError(i18n("Another client is already authenticating, please try again later."));
        result->setCompleted();
        qDebug() << "Another client is already authenticating, please try again later.";
        return;
    }

    m_message = message;
    m_identities = identities;
    m_cookie = cookie;
    m_result = result;
    m_session.clear();

    m_inProgress = true;

    const WId parentId = m_actionsToWID.value(actionId, 0);

    if(!m_view) {
        m_view = new QQuickView;
        m_view->setFlags(Qt::FramelessWindowHint);
        m_view->setWindowState(Qt::WindowFullScreen);
        m_view->setWindowStates(Qt::WindowFullScreen);
        m_view->setColor(QColor(Qt::transparent));
        m_view->rootContext()->setContextProperty("policyKitListener", this);

        KLocalizedString::setApplicationDomain("polkit-kde-agent-1");
        m_view->rootContext()->setContextObject(new KLocalizedContext(m_view));

        m_view->setSource(QUrl::fromLocalFile("/usr/share/jingostest/systemui/org.jingtest.polkit-kde-authentication-agent-1/authdialog.qml"));
        m_view->showFullScreen();

        connect(m_view, &QQuickView::activeChanged, this, &PolicyKitListener::userCancel);
    }

    qDebug() << "WinId of the dialog is " << m_view->winId();
    KWindowSystem::forceActiveWindow(m_view->winId());

    qDebug() << "WinId of the shown dialog is " << m_view->winId();

    m_selectedUser = identities[0];
    m_numTries = 0;
    tryAgain();
}

void PolicyKitListener::tryAgain()
{
    qDebug() << "Trying again";
    m_wasCancelled = false;
    if (m_selectedUser.isValid()) {
        m_session = new Session(m_selectedUser, m_cookie, m_result);
        connect(m_session.data(), SIGNAL(completed(bool)), this, SLOT(completed(bool)));
        m_session.data()->initiate();
    }
}

void PolicyKitListener::finishObtainPrivilege()
{
    qDebug() << "Finishing obtaining privileges";

    if (m_selectedUser.isValid()) {
        m_numTries ++;
    }

    if (!m_gainedAuthorization && !m_wasCancelled && !m_view.isNull()) {
        m_error = i18n("Password input error, please try again");
        setErrorMessage(m_error);
        // if (m_numTries < 3) { //暂时不考虑调用次数的问题
        m_session.data()->deleteLater();
        tryAgain();
        return;
        // }
    }

    if (!m_session.isNull()) {
        m_session.data()->result()->setCompleted();
    } else {
        m_result->setCompleted();
    }

    m_session.data()->deleteLater();
    if (!m_view.isNull()) {
        m_view->hide();
        m_view->deleteLater();
        m_view = nullptr;
    }
    m_inProgress = false;
    qDebug() << "Finish obtain authorization:" << m_gainedAuthorization;
}

bool PolicyKitListener::initiateAuthenticationFinish()
{
    qDebug() << "Finishing authentication";
    return true;
}

void PolicyKitListener::cancelAuthentication()
{
    qDebug() << "Cancelling authentication, please try again";

    m_wasCancelled = true;
    finishObtainPrivilege();
}

void PolicyKitListener::completed(bool gainedAuthorization)
{
    qDebug() << "Completed: " << gainedAuthorization;

    m_gainedAuthorization = gainedAuthorization;

    if(m_gainedAuthorization) {
        invokeSendConfirmSig();
    }
    finishObtainPrivilege();
}
//request
void PolicyKitListener::dialogAccepted(const QString &passwd)
{
    qDebug() << "Dialog accepted";
    if (!m_view.isNull()) {
        m_session.data()->setResponse(passwd);
    }
}

void PolicyKitListener::dialogCanceled()
{
    qDebug() << "Dialog cancelled";

    m_wasCancelled = true;
    if (!m_session.isNull()) {
        m_session.data()->cancel();
    }
    finishObtainPrivilege();
}

void PolicyKitListener::userSelected(const PolkitQt1::Identity &identity)
{
    m_selectedUser = identity;
    // If some user is selected we must destroy existing session
    if (!m_session.isNull()) {
        m_session.data()->deleteLater();
    }
    tryAgain();
}

QStringList  PolicyKitListener::getIdentities() const
{
    QStringList  dentitylist;
    for(int i = 0; i < m_identities.size(); i++) {
        dentitylist << m_identities[i].toString();
    }
     return dentitylist;
}

void PolicyKitListener::setErrorMessage(const QString &errorMessage)
{
    m_error = errorMessage;
    emit errorMessageChanged(errorMessage);
}

QString PolicyKitListener::errorMessage()
{
    return m_error;
}

QString PolicyKitListener::getMessage()
{
    return m_message;
}

void PolicyKitListener::userCancel()
{
    qDebug() << "m_view->isActive()" << m_view->isActive();
    if (m_view == NULL) {
        return;
    }
    if (!m_view->isActive()) {
        dialogCanceled();
    }
}
